### PR TITLE
Extend docs about crawler logic and references implementations

### PR DESCRIPTION
**Pull request by spier**
Hi @Michadelic, 

I had always wanted to add links from your project to the crawler implementations that already exist.
That makes it easier for users to get from using the mock data to using their own project data.

While working in this I was unsure where in your docs to put the documentation about the crawler implementations.

I think it would be best to consolidate all info related to the `repos.json` structure and the crawling in a single place.

I put a quick WIP PR together here, so that you can roughly see what I have in mind.

What I did:
* Consolidating data from `CONTRIBUTING.md` and `README.md` into a new `CRAWLING.md` file that outlines the crawling process. 
* Adding links to reference implementations.

Before I go any further with this I wanted to confirm with you if you would be on board with the general approach.

This still contains a lot of redundant data now etc. Will clean that up when you give me the green light :)

### CODE REVIEW

File: CRAWLING.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+## Reference Implementations
+
+*NOTE*: This section needs to be expanded
+
+You will likely have to adapt all of these crawler implementations to your exact setup. However they may still get you off the ground faster than starting from scratch.
+
+* with jq: <TBD SAMPLE HERE>
+* with ruby: https://github.com/spier/innersource-crawler-ruby
+* with python: https://github.com/zkoppert/innersource-crawler
```
Comment (zkoppert) [inline CRAWLING.md:90]: 🎉

Comment (spier) [reply CRAWLING.md:90]: Oh wow, what dark GitHub magic did you apply @zkoppert to find this so quickly?
Alerts on anything with `https://github.com/zkoppert` or what? :)

**Comment by spier** [comment]
I was thinking about this a bit more, and I have a suspicion now why the instructions for how to get a project listed in the portal were kept in `CONTRIBUTING.md`. 

Is it because when this was a SAP-internal project, getting any repo listed in the portal was considered a "contribution"?

Just trying to understand this so that I can maybe make more helpful proposals to consolidate the documentation.

File: CRAWLING.md (hunk 1 above)
Comment (zkoppert) [reply CRAWLING.md:90]: haha. Nope, I just am subscribed to any changes in this repo. I did that to help me keep track of any patches I need to apply to my deployments of this project.

Comment (spier) [reply CRAWLING.md:90]: Well that actually is a bit more logical than my idea. 
Although I kind of liked the idea of GitHub staff having secret super-power notifications :)

Comment (zkoppert) [reply CRAWLING.md:90]: haha yeah that would be wonderful!

**Comment by Michadelic** [comment]
> I was thinking about this a bit more, and I have a suspicion now why the instructions for how to get a project listed in the portal were kept in `CONTRIBUTING.md`.
> 
> Is it because when this was a SAP-internal project, getting any repo listed in the portal was considered a "contribution"?
> 
> Just trying to understand this so that I can maybe make more helpful proposals to consolidate the documentation.

Hi @spier, you are right, we consider listing a project a tiny contribution, so we kept this info in `CONTRIBUTING.md`. However, as the project matures, i think its better to consolitdate all info about the crawling into a separate file. Will take a look at the code now

**Comment by spier** [comment]
Sounds good @Michadelic. I will push some further commits to this PR later today, with a first stab of how I think the consolidated documentation could look like.

**Comment by spier** [comment]
Hi @Michadelic. I wasn't 100% sure what you meant by "looking at the code" above :)

Could you please confirm again if I should go ahead with this PR and consolidate all info related into `CRAWLING.md`?

**Comment by Michadelic** [comment]
@spier sure, i meant look at the code of this PR, i think it's already in a decent state. I just would keep instrucitons how to list a project separate from how to implement a crawler. What do you think?

**Comment by spier** [comment]
@Michadelic that sounds good.

I think that the issue with "which docs should go where" arises due to different personas. Let's call them **user** and **maintainer**. 

A **user** that wants to get a project listed in the portal only needs the [Listing Project in the Project Portal for InnerSource](https://github.com/SAP/project-portal-for-innersource/blob/main/CONTRIBUTING.md#listing-project-in-the-project-portal-for-innersource) instructions. All further docs about how the portal works from a user's perspective should live in the portal itself, and not in this repo. One could even consider to move the [description section](https://github.com/SAP/project-portal-for-innersource#description) from the README into the portal itself.

The **maintainer** on the other hand needs to know how to install the portal, how to host it, and how to write a customized crawler that works for their org. So those maintainers need all the docs in this repo. The maintainers would also be the ones making upstream contributions to this repo, in case of bugs/features/etc.

## Proposal

Given the thoughts on these two personas, how about about this approach.

### For the **user**

We create `howto-use-this-portal.md` as the docs for the **user**. It would contain:
- the [description section from the README](https://github.com/SAP/project-portal-for-innersource#description)
- the [Listing Project in the Project Portal for InnerSource](https://github.com/SAP/project-portal-for-innersource/blob/main/CONTRIBUTING.md#listing-project-in-the-project-portal-for-innersource) instructions

We can then even link to this file when a user clicks the +-button in the demo portal, instead of the placeholder URL `https://yourcompany.corp/innersource-instructions`.

### For the **maintainer**:

We keep things mostly as they are in the current README, maybe calling out explicitly that this documentation is only required for maintainers.

We extend it with further info about the crawling process. I would start with what I got in `CRAWLING.md` now.
If the README ends up being too long for our liking, we can break it out into a separate file later.

What do you think?

I can create a rough draft of the proposed changes, and push them to this branch. Maybe it is easier to understand if it "feels right" once we can look at something?

**Comment by zkoppert** [comment]
Anything I can do to help here?

**Comment by spier** [comment]
Thanks for the offer @zkoppert. Not that I would know right away.

The documentation approach is relatively clear, I just have to find the time.

As part of this documentation I also wanted to make my crawler implementation public, but maybe that is just scope creep with which I am making the problem too hard for myself. 

Will try to find the time to wrap this up "soon" :)

File: CRAWLING.md [hunk 2]
```diff
@@ -1,14 +1,92 @@
...
-The repository metadata shown in this portal is read from a static `repos.json` file. This project contains a [repos.json](../repos.json) file with mock data for testing and developing purposes.
+The repository metadata shown in this portal is read from a static `repos.json` file. This project contains a [repos.json](./repos.json) file with mock data, which shows the overall structure of that data.
 
-In a productive environment, consider adding an automated crawler script that fetches all InnerSource projects as outlined in the following picture.
+In a productive environment, you will likely need to run your own crawler to populate `repos.json` with the InnerSource projects that you want to show in your portal.
 
-## Step-by-step Crawling Logic
+The picture below shows the relationship between crawler and portal.
+
+In the rest of this document we explain the data structure of `repos.json` and how you would populate it with your own crawler. You will also find [Crawler reference implementations](#reference-implementations) that you can use as starting points for your own crawler.
```
Comment (Michadelic) [inline CRAWLING.md]: suggestion: rest of the document -> following sections

File: CRAWLING.md [hunk 3]
```diff
@@ -1,14 +1,92 @@
 # Crawling
 
-The repository metadata shown in this portal is read from a static `repos.json` file. This project contains a [repos.json](../repos.json) file with mock data for testing and developing purposes.
+The repository metadata shown in this portal is read from a static `repos.json` file. This project contains a [repos.json](./repos.json) file with mock data, which shows the overall structure of that data.
 
-In a productive environment, consider adding an automated crawler script that fetches all InnerSource projects as outlined in the following picture.
+In a productive environment, you will likely need to run your own crawler to populate `repos.json` with the InnerSource projects that you want to show in your portal.
 
-## Step-by-step Crawling Logic
+The picture below shows the relationship between crawler and portal.
```
Comment (Michadelic) [inline CRAWLING.md]: Suggestion to add the hint about GitHub here as it is shown in the graphic (?): The picture below shows the relationship between crawler, portal, and repositories. We assume that the repositories are located in GitHub and use the GitHub API to fetch repository metadata. If you are using a different source code management system you can fill the structure described below with the available data.

File: CRAWLING.md [hunk 4]
```diff
@@ -1,14 +1,92 @@
...
+
+ > *Note:* This assumes that all projects that you want to display have the topic `inner-source`. You can of course use any other topic as well. You can optionally limit the results further  by adding `is:public` or `is:private` to the query, depending on how InnerSource repositories are characterized in your environment.
+
+### Step 2: (Optional) Extended Project Information from GitHub
+
+For a richer portal experience, you can add extended project information that is available via further GitHub API calls.
+
+For each project retrieved in step (1), add a key `_InnerSourceMetadata` with an object containing the following metadata about the project:
+
+* `topics` (Optional): Query GitHub [topics](https://docs.github.com/en/rest/reference/repos#get-all-repository-topics) and add the array of topics with the key `topics` for each repo to allow searching projects by topic and displaying them on the detail popup.
```
Comment (Michadelic) [inline CRAWLING.md]: since it is now all optional we can remove the (Optional) in the list of items

File: CRAWLING.md [hunk 5]
```diff
@@ -1,14 +1,92 @@
...
+* `participation`: Query GitHub for the [weekly commit count](https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#get-the-weekly-commit-count) (subset "all") and add it. with the key `participation`
+* `guidelines` (Optional): Check if there are contribution guidelines and add the file name with the key `guidelines` (e.g. `CONTRIBUTING.md`). If specified, the *Contribute* button for this project will link to this file instead of the repository root.
```
Comment (Michadelic) [inline CRAWLING.md]: same here

File: CRAWLING.md [hunk 6]
```diff
@@ -1,14 +1,92 @@
...
+
+You will have to adapt all of these crawler implementations to your exact setup. However they may give you a good starting points.
+
+* with curl/jq: TBD SAMPLE HERE
+* with ruby: [spier/innersource-crawler-ruby](https://github.com/spier/innersource-crawler-ruby)
+* with python: [zkoppert/innersource-crawler](https://github.com/zkoppert/innersource-crawler)
+
+
+
+# BACKUP
```
Comment (Michadelic) [inline CRAWLING.md]: can be safely removed now as the new version is so much nicer

**Review by Michadelic — APPROVED** [review]
Awesome, it reads really well now with the three step apporach.
I left some comments and suggestions in the review, let me know what you think.

I would also add a note about implementing crawlers for other source code systems as a call to action to the Contributing.md - can be done in a separate commit later on.

File: docs/CRAWLING.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+You can customize the portal further with the following properties, that you can add to the `_InnerSourceMetadata` object explained in step (2):
+
+* `score` (Optional): Calculate the [Repository Activity Score](https://patterns.innersourcecommons.org/p/repository-activity-score) to define a meaningful order for the projects. Add it with the key `score`.
+* Check if there is a file `innersource.json` in the repository and add all keys from that file directly below `_InnerSourceMetadata`. For further details review the [syntax definition](LISTING.md#syntax-definition-of-innersourcejson) of `innersource.json`.
+
+## Reference Implementations
+
+You will have to adapt all of these crawler implementations to your exact setup. However they may give you a good starting points.
+
+* with curl/jq: TBD SAMPLE HERE
```
Comment (Michadelic) [inline docs/CRAWLING.md:78]: would be great to have the 3 minute example here, i played around with it yesterday but did not get it working immediately

**Comment by spier** [comment]
This PR is ready for a proper review now.

Open questions are:

- how to point from README to LISTING?
- in CRAWLING steps 2 and 3 are optional. Some of the bullets below those steps are marked as optional too. That seems redundant. We should validate that this data is indeed optional and that the portal works without them
- CRAWLING mentions that "You can write the API response to repos.json as is". That might not be 100% correct. Maybe we can fix that when adding a curl/jq example of the most minimal crawling approach :)

Looking forward to your feedback.

File: CRAWLING.md (hunk 6 above)
Comment (spier) [reply CRAWLING.md]: Done

**Review by Michadelic — APPROVED** [review]
The PR is now in a very good state, thank you very much!

I added some suggestions and ideas that can be added on top, some questions from above are also answered:
* new section to point to LISTING.md
* removing (optional) for the bullets, the features of step 2+3 should all work independently
* I tested yesterday with the pure GitHub response (removed all steps that enrich the GitHub metadata in our crawler) so it worked fine. To be verified with the minimal CURL example, of course

Do you want me to create a commit on top of this PR for the suggestions or should we merge as is and take it from there?

**Review by Michadelic — APPROVED** [review]
looks good to me

**Comment by Michadelic** [comment]
thank you very much for this great contribution, the documentation is now much more readable!

**Pull request merged**